hypothesis-python/tests/cover/test_stateful.py

```diff
@@ -906,52 +906,49 @@ def fail_fast(self, param):
 
 
 def test_initialize_rule_dont_mix_with_precondition():
-    with pytest.raises(InvalidDefinition) as exc:
+    match = "An initialization rule cannot have a precondition."
+    with pytest.raises(InvalidDefinition, match=match):
 
         class BadStateMachine(RuleBasedStateMachine):
             @precondition(lambda self: True)
             @initialize()
             def initialize(self):
```

Comment:
If we're raising at definition-time, it seems useful to point to the qualname of the rule in question, and then to test for that too.

Reply:
Seems like we might as well do this everywhere - except `f.__qualname__` includes `<locals>` since we define the rules inside a nested function, so I split on that before printing.